packages/fetchai/connections/p2p_libp2p/aea/dhtpeer/dhtpeer.go

```diff
@@ -0,0 +1,818 @@
+				dhtPeer.publicKey, ":", err.Error())
+			return
+		}
+		sPeerID = peerID.Pretty()
+	} else {
+		// needed when a relay client queries for a peer ID
+		log.Println("DEBUG did NOT found the address locally, looking for it in the DHT...")
+		peerID, err := dhtPeer.lookupAddressDHT(reqAddress)
+		if err == nil {
+			log.Println("DEBUG found address", reqAddress, "on the DHT")
+			sPeerID = peerID.Pretty()
+		} else {
+			log.Println("ERROR did NOT find address", reqAddress, " locally or on the DHT.")
+			return
+		}
+	}
+
+	log.Println("DEBUG sending peer id", sPeerID, "for address", reqAddress)
+	err = writeBytes(stream, []byte(sPeerID))
+	if err != nil {
+		log.Println("ERROR While sending peerID to peer:", err)
+	}
+}
+
+func (dhtPeer *DHTPeer) handleAeaNotifStream(stream network.Stream) {
+	log.Println("DEBUG Got a new notif stream")
+
+	if !dhtPeer.addressAnnounced {
+		if dhtPeer.myAgentAddress != "" {
+			err := dhtPeer.registerAgentAddress(dhtPeer.myAgentAddress)
+			if err != nil {
+				log.Println("ERROR while announcing my agent address to dht:" + err.Error())
+				return
+			}
+		}
+		if dhtPeer.enableRelay {
+			for addr := range dhtPeer.dhtAddresses {
+				err := dhtPeer.registerAgentAddress(addr)
+				if err != nil {
+					log.Println("ERROR while announcing relay client address", addr, ":", err)
+				}
+			}
+
+		}
+		if dhtPeer.delegatePort != 0 {
+			for addr := range dhtPeer.tcpAddresses {
+				err := dhtPeer.registerAgentAddress(addr)
+				if err != nil {
+					log.Println("ERROR while announcing delegate client address", addr, ":", err)
+				}
+			}
+
+		}
+	}
+	dhtPeer.addressAnnounced = true
+}
+
+func (dhtPeer *DHTPeer) handleAeaRegisterStream(stream network.Stream) {
+	log.Println("DEBUG Got a new aea register stream")
+
+	clientAddr, err := readBytes(stream)
+	if err != nil {
+		log.Println("ERROR While reading client Address from stream:", err)
+		err = stream.Reset()
+		ignore(err)
+		return
+	}
+
+	err = writeBytes(stream, []byte("doneAddress"))
+	ignore(err)
+
+	clientPeerID, err := readBytes(stream)
+	if err != nil {
+		log.Println("ERROR While reading client peerID from stream:", err)
+		err = stream.Reset()
+		ignore(err)
+		return
+	}
+
+	err = writeBytes(stream, []byte("donePeerID"))
+	ignore(err)
+
+	log.Println("DEBUG Received address registration request (addr, peerid):", clientAddr, clientPeerID)
+	dhtPeer.dhtAddresses[string(clientAddr)] = string(clientPeerID)
+	if dhtPeer.addressAnnounced {
+		log.Println("DEBUG Announcing client address", clientAddr, clientPeerID, "...")
+		err = dhtPeer.registerAgentAddress(string(clientAddr))
+		if err != nil {
+			log.Println("ERROR While announcing client address", clientAddr, "to the dht:", err)
+			err = stream.Reset()
+			ignore(err)
+			return
+		}
+	}
+}
+
+func (dhtPeer *DHTPeer) registerAgentAddress(addr string) error {
+	addressCID, err := computeCID(addr)
+	if err != nil {
+		return err
+	}
+
+	// TOFIX(LR) tune timeout
```

Comment:
@lrahmani what do you have in mind?

Reply:
I didn't give it much thoughts, but considering profiling/benchmarks as a first approximation.
a More appropriate solution is to attempt an implementation with channels, but I need to investigate that